### PR TITLE
Enable using OpenZeppelin Platform for deployments

### PR DESCRIPTION
**Pull request by ericglau**
Remaining:
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: packages/core/scripts/copy-build-info.js [hunk 1]
```diff
@@ -47,10 +47,48 @@ assert(
...
+assert(origSelection.includes('evm.bytecode'));
+assert(origSelection.includes('evm.deployedBytecode'));
+assert(origSelection.includes('evm.methodIdentifiers'));
+assert(origSelection.includes('metadata'));
+
+input.settings.outputSelection['*'] = {
+  '*': ['abi', 'evm.bytecode', 'evm.deployedBytecode', 'evm.methodIdentifiers', 'metadata'],
+};
+
+// Keep only relevant sections of build info
```
Comment (frangio) [inline packages/core/scripts/copy-build-info.js:79]: What do we intend to remove? Can we specify that in the comment?

If I understand correctly we're removing only from the `output` object and keeping only `output.contracts` as filtered above, but also removing the AST for example.

I assume this is to reduce the package size? We could also specify that in a comment somewhere.

File: packages/core/scripts/copy-build-info.js [hunk 2]
```diff
@@ -47,10 +47,48 @@ assert(
...
+buildInfo = {
+  _format: buildInfo._format,
```
Comment (frangio) [inline packages/core/scripts/copy-build-info.js:81]: I'm (very slightly) worried that we're modifying the file structure but preserving the `_format` string as is. Perhaps we should change it to like `${buildInfo._format}-slim1`.

File: packages/core/scripts/copy-build-info.js [hunk 3]
```diff
@@ -47,10 +47,48 @@ assert(
...
+  for (const contractName in contractNames) {
+    contractNames[contractName] = {
+      abi: contractNames[contractName].abi,
+      evm: contractNames[contractName].evm,
+      metadata: contractNames[contractName].metadata, // metadata is needed to determine the license type
+    };
+  }
+}
+
+// Adjust correponding output selection
```
Comment (frangio) [inline packages/core/scripts/copy-build-info.js:65]: This is clever... but we should test that it actually works and that the end result can be verified. I don't trust the Solidity compiler to produce exactly the same bytecode with the modified settings. Perhaps we should: 1) check with the Solidity team if they think this should work, and 2) manually check that we can verify contracts like this (you may have already tried?).

File: packages/core/src/impl-store.ts [hunk 4]
```diff
@@ -36,6 +43,7 @@ async function fetchOrDeployGeneric<T extends Deployment, U extends T = T>(
   deploy: () => Promise<U>,
   opts?: DeployOpts,
   merge?: boolean,
+  getDeploymentResponse?: (deploymentId: string, allowUndefined: boolean) => Promise<DeploymentResponse | undefined>,
```
Comment (frangio) [inline packages/core/src/impl-store.ts:46]: This new param is missing documentation. Same for the other functions.

File: packages/core/src/impl-store.ts [hunk 5]
```diff
@@ -180,11 +205,18 @@ async function checkForAddressClash(
       debug('deleting a previous deployment at address', updated.address);
       clash.set(undefined);
     } else {
-      throw new Error(
-        `The following deployment clashes with an existing one at ${updated.address}\n\n` +
-          JSON.stringify(updated, null, 2) +
-          `\n\n`,
-      );
+      const existing = clash.get();
+      // it's a clash if there is no deployment id or if deployment ids don't match
```
Comment (frangio) [inline packages/core/src/impl-store.ts:209]: Can you describe for me in what scenario we would see this? Finding an existing entry with the same deployment id.

**Review by frangio — COMMENTED** [review]
Very early comments, still a lot to go through.

File: docs/modules/ROOT/pages/platform-deploy.adoc [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  }
+}
+----
+
+[[usage]]
+== Usage
+
+When using the xref:api-hardhat-upgrades.adoc[Hardhat Upgrades API functions], enable OpenZeppelin Platform deployments using any of the ways below.
+
+NOTE: Only functions that have the `platform` option in their API reference support deployments through the OpenZeppelin Platform. If you enable the following but use functions that do not support `platform`, those functions will deploy using ethers.js.
```
[the remark this answers is not in the document]
Comment (ericglau) [reply docs/modules/ROOT/pages/platform-deploy.adoc:25]: Scripts are meant to be reusable between a local dev network, testnets, and production.  The recommendation would be to set `deploy: true` in the config when using testnets/production (where platform will be used), so that the scripts don't need to be modified from when using a local network.

So there could be two modes:
- If using `platform.<function>` or using the option `{platform: true}`, then give an error if the function does not support platform.
- If using `deploy: true`, then just fall back if the function does not support platform, as described in the note.  If the function does support platform, but platform is not available (e.g. API key is not set), it would still give an error.

File: packages/core/src/impl-store.ts (hunk 5 above)
Comment (ericglau) [reply packages/core/src/impl-store.ts:209]: When calling resumeOrDeploy before checkForAddressClash, resumeOrDeploy may detect that the transaction hash changed for the deployment id.  In this case, it returns an updated entry for the deployment.

We then call checkForAddressClash with this updated entry.  The address and deployment ids match, but the transaction hash changed, so we don't want this to be treated as a clash.  We just want to update the manifest.

I'll add a test for this.

Comment (ericglau) [reply packages/core/src/impl-store.ts:209]: Added test in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/763/commits/a9b5147abdd3256d974ff2a803ae5302232c0903

File: packages/core/scripts/copy-build-info.js (hunk 3 above)
Comment (ericglau) [reply packages/core/scripts/copy-build-info.js:65]: I was able to verify a contract using this.

Would it be safer (although a bit incorrect) to keep the original output selection here, even if we remove it from the actual output?

File: packages/plugin-hardhat/src/index.ts [hunk 7]
```diff
@@ -157,6 +132,71 @@ if (tryRequire('@nomiclabs/hardhat-etherscan')) {
...
+    beacon: {
+      getImplementationAddress: (beaconAddress: string) =>
+        getImplementationAddressFromBeacon(hre.network.provider, beaconAddress),
+    },
+    proposeUpgrade: makeProposeUpgrade(hre, platform),
+    verifyDeployment: makeVerifyDeploy(hre),
+    verifyDeploymentWithUploadedArtifact: makeVerifyDeployWithUploadedArtifact(hre),
+    getDeploymentArtifact: makeGetVerifyDeployArtifact(hre),
+    getDeploymentBuildInfo: makeGetVerifyDeployBuildInfo(hre),
+    getBytecodeDigest: makeGetBytecodeDigest(hre),
```
[the remark this answers is not in the document]
Comment (ericglau) [reply packages/plugin-hardhat/src/index.ts:196]: These are cloned from the `hardhat-defender` package.  We could make these only available for `platform.<fn>` and not `upgrades.<fn>`

However, `deployContract` and `proposeUpgrade` should be considered separately since they take the `platform` option.

`upgrades.deployContract` without `platform` option is not allowed anyways and would fail.  While the syntax `upgrades.deployContract(...{platform: true})` would be consistent with what we allow for `deployProxy`, I no longer see any real benefit to allowing it for this case (other than having one less import/require in the user's code).

`upgrades.proposeUpgrade` without `platform` option could make sense.  This would mean deploy the new implementation using ethers.js, then submit the proposal on Platform (even though Platform is not part of the name).

---
To summarize:

Support:
```
platform.deployContract
platform.verifyDeployment
platform.verifyDeploymentWithUploadedArtifact
platform.getDeploymentArtifact
platform.getDeploymentBuildInfo
platform.getBytecodeDigest
platform.proposeUpgrade
upgrades.proposeUpgrade
```

Do not support:
```
upgrades.deployContract
upgrades.verifyDeployment
upgrades.verifyDeploymentWithUploadedArtifact
upgrades.getDeploymentArtifact
upgrades.getDeploymentBuildInfo
upgrades.getBytecodeDigest
```

Let me know if you agree.

File: docs/modules/ROOT/pages/platform-deploy.adoc [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    apiKey: process.env.API_KEY,
+    apiSecret: process.env.API_SECRET,
+  }
+}
+----
+
+[[usage]]
+== Usage
+
+When using the xref:api-hardhat-upgrades.adoc[Hardhat Upgrades API functions], enable OpenZeppelin Platform deployments using any of the ways below.
```
[the remark this answers is not in the document]
Comment (ericglau) [reply docs/modules/ROOT/pages/platform-deploy.adoc:23]: Let me know what you think of this:
- If you want to use the same scripts on local devnets and Platform, use `upgrades.<fn>` with `deploy: true` in the config that you use with Platform.
- If you only run your scripts on Platform, use `platform.<fn>` for clarity.

File: packages/core/scripts/copy-build-info.js (hunk 2 above)
Comment (ericglau) [reply packages/core/scripts/copy-build-info.js:81]: Any benefit to include an `oz` marker for easier identification? `${buildInfo._format}-oz-slim1`

File: docs/modules/ROOT/pages/platform-deploy.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (frangio) [reply docs/modules/ROOT/pages/platform-deploy.adoc:25]: That sounds reasonable.

File: docs/modules/ROOT/pages/platform-deploy.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (frangio) [reply docs/modules/ROOT/pages/platform-deploy.adoc:23]: I would say to use `platform.<fn>` when the user wants to make sure Platform is used and wants to see an error otherwise, but mostly they should use the first option you describe.

File: packages/plugin-hardhat/src/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (frangio) [reply packages/plugin-hardhat/src/index.ts:196]: What do you think of putting `platform` under `upgrades`?
So we would have
```
upgrades: {
  deployProxy: ...,
  ...,
  platform: {
    deployProxy: ..., // aliased with { platform: true }
    deployContract: ..., // only at this level
    ...
  }
}
```

File: packages/core/scripts/copy-build-info.js (hunk 3 above)
Comment (frangio) [reply packages/core/scripts/copy-build-info.js:65]: Yeah I'd feel more confident in that. I think that + using the `-slim` format would be good.

File: packages/core/scripts/copy-build-info.js (hunk 2 above)
Comment (frangio) [reply packages/core/scripts/copy-build-info.js:81]: Sounds good too.

File: packages/core/src/deployment.ts [hunk 9]
```diff
@@ -94,58 +101,113 @@ async function validateCached<T extends Deployment>(
...
+    const existingBytecode = await getCode(provider, deployment.address);
     if (isEmpty(existingBytecode)) {
-      throw new InvalidDeployment(stored);
+      throw new InvalidDeployment(deployment);
     }
   }
+  return deployment;
+}
+
+export interface DeploymentResponse {
```
Comment (frangio) [inline packages/core/src/deployment.ts:161]: "Deployment response" doesn't seem like the right concept to me... I think it should be either `DeploymentStatus` or something like `RemoteDeployment` or even `PlatformDeployment`. This would apply as well to the function names.

I'm thinking we should go with the latter and use `remoteDeploymentId` in the manifest. `deploymentId` is very generic.

File: packages/core/src/impl-store.ts (hunk 4 above)
Comment (frangio) [reply packages/core/src/impl-store.ts:46]: I realize now that the entire function is missing documentation. :stuck_out_tongue: That's on me. Could be a good moment to add some though.

Comment (frangio) [reply packages/core/src/impl-store.ts:46]: I realize now that the entire function is missing documentation. :stuck_out_tongue: That's on me. Could be a good moment to add some though.

File: packages/core/src/deployment.ts [hunk 10]
```diff
@@ -217,3 +281,38 @@ export class InvalidDeployment extends Error {
...
+export async function isDeploymentCompleted(
+  address: string,
+  deploymentId: string,
+  getDeploymentResponse: (allowUndefined: boolean) => Promise<DeploymentResponse | undefined>,
+): Promise<boolean> {
+  debug('verifying deployment id', deploymentId);
+  const response = await getDeploymentResponse(false);
+  if (response === undefined) {
+    throw new Error(`Broken invariant: Response not found for deployment id ${deploymentId}`);
+  }
```
Comment (frangio) [inline packages/core/src/deployment.ts:303]: It seems like the only place where we use `allowUndefined = false` we check the result anyway. I would simplify `getDeploymentResponse` removing the `allowUndefined` argument, making it always return `undefined` if the deployment is not found.

In this function I would also make `getDeploymentResponse` a function that takes a deployment id, instead of hardcoding it at the callsite.

I think both of these changes should make invocation simpler because you would just pass along the `getDeploymentResponse` function.

File: packages/plugin-hardhat/contracts/DeployContract.sol [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+contract NonUpgradeable {
+}
+
+contract IsInitializable is Initializable1 {
+}
+
+contract IsInitializableUpgradeable is Initializable2 {
+}
+
+contract IsUUPS is UUPSUpgradeable {
```
Comment (frangio) [inline packages/plugin-hardhat/contracts/DeployContract.sol:17]:
```suggestion
// Insecure contract
contract IsUUPS is UUPSUpgradeable {
```

File: packages/core/src/deployment.ts [hunk 12]
```diff
@@ -22,6 +22,10 @@ export interface Deployment {
   txHash?: string;
 }
 
+export interface DeploymentId {
+  deploymentId?: string;
```
Comment (frangio) [inline packages/core/src/deployment.ts:26]: What's the reason not to add this directly in `Deployment`?

File: packages/plugin-hardhat/src/utils/options.ts [hunk 13]
```diff
@@ -32,16 +32,29 @@ type Initializer = {
...
-export type DeployProxyOptions = StandaloneOptions & Initializer;
+export type Platform = {
+  platform?: boolean;
+};
+
+export type PlatformSupportedOptions = Platform & {
+  verifySourceCode?: boolean;
+};
+
+export type DeployBeaconProxyOptions = DeployOpts & ProxyKindOption & Initializer & PlatformSupportedOptions;
```
Comment (frangio) [inline packages/plugin-hardhat/src/utils/options.ts:43]: What's the reason `DeployOpts` was added here but not in all `Deploy*Options` types?

File: packages/plugin-hardhat/src/prepare-upgrade.ts [hunk 14]
```diff
@@ -23,8 +23,12 @@ export type PrepareUpgradeFunction = (
   opts?: PrepareUpgradeOptions,
 ) => Promise<DeployImplementationResponse>;
 
-export function makePrepareUpgrade(hre: HardhatRuntimeEnvironment): PrepareUpgradeFunction {
+export function makePrepareUpgrade(hre: HardhatRuntimeEnvironment, platformModule: boolean): PrepareUpgradeFunction {
   return async function prepareUpgrade(proxyOrBeacon, ImplFactory, opts: PrepareUpgradeOptions = {}) {
+    if (platformModule && opts.platform === undefined) {
+      opts.platform = true;
+    }
```
Comment (frangio) [inline packages/plugin-hardhat/src/prepare-upgrade.ts:30]: Similar comment about mutation here. Plus let's have an `enablePlatform` helper?

File: packages/plugin-hardhat/src/type-extensions.ts [hunk 15]
```diff
@@ -5,5 +5,22 @@ import type { HardhatUpgrades } from '.';
...
   export interface HardhatRuntimeEnvironment {
     upgrades: HardhatUpgrades;
+    platform: HardhatUpgrades;
+  }
+}
+
+export interface HardhatPlatformConfig {
+  apiKey: string;
+  apiSecret: string;
+  deploy?: boolean;
```
Comment (frangio) [inline packages/plugin-hardhat/src/type-extensions.ts:15]: This option name is a little confusing I think. Perhaps we should do `useDeploy` or `preferDeploy`.

File: packages/plugin-hardhat/src/deploy-proxy.ts [hunk 16]
```diff
@@ -73,6 +84,13 @@ export function makeDeployProxy(hre: HardhatRuntimeEnvironment): DeployFunction
     const inst = ImplFactory.attach(proxyDeployment.address);
     // @ts-ignore Won't be readonly because inst was created through attach.
     inst.deployTransaction = proxyDeployment.deployTransaction;
+    if (opts.platform && proxyDeployment.deploymentId !== undefined) {
+      inst.deployed = async () => {
+        assert(proxyDeployment.deploymentId !== undefined);
+        await waitForDeployment(hre, opts, inst.address, proxyDeployment.deploymentId);
+        return inst;
```
Comment (frangio) [inline packages/plugin-hardhat/src/deploy-proxy.ts:91]: Is there a chance we might want to update `inst.deployTransaction` here?

Should we also add a call to the original `inst.deployed()` at the end?

File: packages/plugin-hardhat/src/deploy-beacon-proxy.ts [hunk 17]
```diff
@@ -72,16 +80,23 @@ export function makeDeployBeaconProxy(hre: HardhatRuntimeEnvironment): DeployBea
...
     const inst = attachTo.attach(proxyDeployment.address);
     // @ts-ignore Won't be readonly because inst was created through attach.
     inst.deployTransaction = proxyDeployment.deployTransaction;
+    if (opts.platform && proxyDeployment.deploymentId !== undefined) {
+      inst.deployed = async () => {
+        assert(proxyDeployment.deploymentId !== undefined);
+        await waitForDeployment(hre, opts, inst.address, proxyDeployment.deploymentId);
+        return inst;
+      };
+    }
```
Comment (frangio) [inline packages/plugin-hardhat/src/deploy-beacon-proxy.ts:99]: Same here regarding updating `deployTransaction`. We might want to define a helper for this entire piece of code.

File: packages/plugin-hardhat/src/platform/verify-deployment.ts
Comment (frangio) [inline packages/plugin-hardhat/src/platform/verify-deployment.ts:1]: What is the difference between `verifyDeploy` and `verifyDeployWithUploadedArtifact`? Do we need both? Maybe we can merge them.

I couldn't find docs for this file and I'm not sure what to make of this functionality.

File: packages/plugin-hardhat/src/platform/deploy.ts [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+  const client = getPlatformClient(hre);
+
+  const constructorArgs = [...args] as (string | number | boolean)[];
+  const contractInfo = await getContractInfo(hre, factory, { constructorArgs, ...opts });
+  const network = await getNetwork(hre);
+  debug(`Network ${network}`);
+
+  const verifySourceCode = opts.verifySourceCode ?? true;
+  if (verifySourceCode) {
+    await registerEtherscanApiKey(hre, network, client.BlockExplorerApiKey);
```
Comment (frangio) [inline packages/plugin-hardhat/src/platform/deploy.ts:67]: As discussed offline we may want to reconsider doing this automatically.

File: packages/plugin-hardhat/src/platform/deploy.ts [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+    for (const key of registeredKeys) {
+      if (key.network === network) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+function getContractPathAndName(fullyQualified: string) {
```
Comment (frangio) [inline packages/plugin-hardhat/src/platform/deploy.ts:127]:
```ts
import { parseFullyQualifiedName } from "hardhat/utils/contract-names";
```

File: packages/plugin-hardhat/src/platform/deploy.ts [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+
+  const metadata = JSON.parse(metadataString);
+
+  const license = metadata.sources[contractInfo.contractPath].license;
+  if (license === undefined) {
+    debug('License not found in metadata');
+    return undefined;
+  }
+
+  debug(`Found license from metadata: ${license}`);
```
Comment (frangio) [inline packages/plugin-hardhat/src/platform/deploy.ts:192]:
```suggestion
  if (license === undefined) {
    debug('License not found in metadata');
  } else {
    debug(`Found license from metadata: ${license}`);
  }
```

File: packages/plugin-hardhat/src/platform/propose-upgrade.ts [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+    } else {
+      // try getting the implementation address so that it will give an error if it's not a transparent/uups proxy
+      await getImplementationAddress(hre.network.provider, proxyAddress);
+    }
+
+    const implFactory =
+      typeof contractNameOrImplFactory === 'string'
+        ? await hre.ethers.getContractFactory(contractNameOrImplFactory)
+        : contractNameOrImplFactory;
+    const contractName = typeof contractNameOrImplFactory === 'string' ? contractNameOrImplFactory : undefined;
```
Comment (frangio) [inline packages/plugin-hardhat/src/platform/propose-upgrade.ts:62]: So if you want verification you need to provide the contract name as opposed to the impl factory?

File: packages/plugin-hardhat/src/platform/propose-upgrade.ts [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+
+    let txResponse;
+    let newImplementation: string;
+
+    if (typeof prepareUpgradeResult === 'string') {
+      newImplementation = prepareUpgradeResult;
+    } else {
+      txResponse = prepareUpgradeResult;
+      // The txResponse is a create2 transaction, so the deployment address is not easily retrievable from it.
+      // Instead, run the prepareUpgrade again but this time adjust the options to request the address and ensure it does not attempt another deployment.
```
Comment (frangio) [inline packages/plugin-hardhat/src/platform/propose-upgrade.ts:78]: This seems hackish, do we have other options?

File: packages/plugin-hardhat/src/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ericglau) [reply packages/plugin-hardhat/src/index.ts]: Removed everything related to `verifyDeployment` since that is handled automatically by the Platform deploy API.

I don't think `upgrades.platform.deployContract` makes sense since it is unrelated to upgradeable contracts.

Now since only `deployContract` and `proposeUpgrade` are new, I changed it to this:
```
upgrades: {
  deployProxy: ...,
  prepareUpgrade: ...,
  ...,
},
platform: {
  deployProxy: ..., // aliased with { platform: true }
  prepareUpgrade: ..., // aliased with { platform: true }
  ...,
  deployContract: ..., // only at this level
}
```

File: packages/plugin-hardhat/src/prepare-upgrade.ts (hunk 14 above)
Comment (ericglau) [reply packages/plugin-hardhat/src/prepare-upgrade.ts]: The helper is named as `withPlatformDefaults`

File: packages/plugin-hardhat/src/platform/propose-upgrade.ts (hunk 21 above)
Comment (ericglau) [reply packages/plugin-hardhat/src/platform/propose-upgrade.ts]: Removed this since verification is handled automatically by Deploy API

File: packages/plugin-hardhat/src/platform/verify-deployment.ts
Comment (ericglau) [reply packages/plugin-hardhat/src/platform/verify-deployment.ts:1]: Removed since this is handled automatically by Deploy API

File: packages/plugin-hardhat/src/platform/deploy.ts (hunk 18 above)
Comment (ericglau) [reply packages/plugin-hardhat/src/platform/deploy.ts]: Removed since this is expected to be handled automatically by the Platform

File: packages/core/src/deployment.ts (hunk 12 above)
Comment (ericglau) [reply packages/core/src/deployment.ts]: There are some cases where we use `Required<Deployment>` to enforce that a deployment has a `txHash` defined.

For example, `deployProxy` previously used `Required<ProxyDeployment & DeployTransaction>` where `ProxyDeployment` extends `Deployment`.

We want the `deploymentId` to be optional, because it is only relevant when using Platform deploys.  So it cannot be part of `Deployment` if the above pattern continues to be used.

So now, `deployProxy` can use `Required<ProxyDeployment & DeployTransaction> & RemoteDeploymentId;`

File: packages/plugin-hardhat/src/utils/options.ts (hunk 13 above)
Comment (ericglau) [reply packages/plugin-hardhat/src/utils/options.ts:58]: The other `Deploy*Options` types inherit `StandaloneOptions` which already has `DeployOpts`.  Added doc comments to clarify some of the inherited options.

File: docs/modules/ROOT/pages/platform-deploy.adoc [hunk 23]
```diff
@@ -0,0 +1,87 @@
+= Using with OpenZeppelin Platform
+
+The Hardhat Upgrades package can use the OpenZeppelin Platform for deployments instead of ethers.js.
```
[the remark this answers is not in the document]
Comment (ericglau) [reply docs/modules/ROOT/pages/platform-deploy.adoc]: Added a brief note but we should plan to improve this when there is more documentation to link to.

File: docs/modules/ROOT/pages/platform-deploy.adoc [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+
+[[configuration]]
+== Configuration
+
+Create a deployment environment on the OpenZeppelin Platform and provide the Team API Key and secret in your `hardhat.config.js` or `hardhat.config.ts` file under `platform`:
```
[the remark this answers is not in the document]
Comment (ericglau) [reply docs/modules/ROOT/pages/platform-deploy.adoc:8]: We should plan to improve this when there is more documentation to link to

File: packages/plugin-hardhat/src/platform/propose-upgrade.ts (hunk 22 above)
Comment (ericglau) [reply packages/plugin-hardhat/src/platform/propose-upgrade.ts]: Improved to use the deployed impl object (containing both txResponse and impl address) directly.

File: packages/plugin-hardhat/src/utils/deploy-impl.ts [hunk 25]
```diff
@@ -116,17 +116,15 @@ async function deployImpl(
...
+    remoteDeploymentId => getRemoteDeployment(hre, remoteDeploymentId),
   );
 
   let txResponse;
   if (opts.getTxResponse) {
     if ('deployTransaction' in deployment) {
       txResponse = deployment.deployTransaction;
     } else if (deployment.txHash !== undefined) {
-      txResponse = hre.ethers.provider.getTransaction(deployment.txHash);
+      txResponse = await hre.ethers.provider.getTransaction(deployment.txHash);
```
Comment (frangio) [inline packages/plugin-hardhat/src/utils/deploy-impl.ts:127]: This error was possible because this function returns `Promise<any>`. A cautionary tale! Is there any chance we could get rid of that `any`?

**Review by frangio — APPROVED** [review]
Changes look good!
